### PR TITLE
Add Write support to OCP/ICalendar

### LINKED ISSUE #20154: PHP Calendar API: Write support
`OCP\Calendar\IManager` is currently only providing a read-only API.
This is drastically limiting the number of use-cases, forcing app-developers to use the CalDAV interface via javascript, if they desire write-support for the user's calendar.

Nextcloud should provide a PHP api with write-support for the user's calendar.

Due to it's simple design, our current API is limited to read-only. Adding write-support to the existing API is not possible. This is mostly caused by the structure of events.
They are returned as a very simple key-value array. Some information from the ICS is already lost when the conversion takes place. Converting modified key-value arrays back to ICS could possibly lead to data-loss.

---

Proposal:
1. Deprecate `\OCP\Calendar\IManager` and `\OCP\Calendar\ICalendar` with Nextcloud 20 and remove with Nextcloud 23.

2.
- `Implement` `\OCP\Calendar\IManagerV2`:
```php
interface \OCP\Calendar\IManagerV2{
        public function isEnabled():bool;
        public function getCalendars():ICalendarV2[];
        public function getCalendar(string $key):?ICalendarV2;
        public function search(string $pattern, array $searchProperties=[], array $options=[], int $limit=null, int $offset=null)?ICalendarObjectV2[];
        public function registerCalendar(ICalendarV2 $calendar):void;
        public function unregisterCalendar(ICalendarV2 $calendar):void;
        public function register(\Closure $callable):void;
        public function clear():void;
}
```

- `Implement` `\OCP\Calendar\ICalendarV2`:
```php
interface \OCP\Calendar\ICalendarV2{
	public function getKey():string;
	public function getDisplayName():?string;
	public function getDisplayColor():?string;
        public function isWriteable():bool;
        public function getByUri(string $uri):?ICalendarObjectV2;
        public function search(string $pattern, array $searchProperties=[], array $options=[], int $limit=null, int $offset=null)?ICalendarObjectV2[];
        public function create(\Sabre\VObject\VCalendar $vObject):ICalendarObjectV2;
}
```

- `Implement` `\OCP\Calendar\ICalendarObjectV2`:
```php
interface \OCP\Calendar\ICalendarObjectV2{
        public function getCalendarKey():string;
        public function getUri():string;
        public function getVObject():\Sabre\VObject\VCalendar;
        public function update():void;
        public function delete():void;
}
```

### PR DESCRIPTION
**Pull request by rotdrop**
See issue #20154. Fixes #20154. Updates #20178.

### CODE REVIEW

File: apps/dav/lib/AppInfo/Application.php [hunk 1]
```diff
@@ -123,6 +125,8 @@ public function registerHooks(HookManager $hm,
 								   EventDispatcherInterface $dispatcher,
 								   IAppContainer $container,
 								   IServerContainer $serverContainer) {
+		/** @var HookManager $hm */
+		$hm = $this->getContainer()->query(HookManager::class);
```
Comment (ChristophWurst) [inline apps/dav/lib/AppInfo/Application.php:129]: you get this as argument already

Comment (rotdrop) [reply apps/dav/lib/AppInfo/Application.php]: b20657ee317160135dc4de6ad15d0e1f4bc38c9e

**Comment by tcitworld** [comment]
There's been a lot of changes since then, especially https://github.com/nextcloud/server/pull/29188, so it needs a rebase.

**Comment by skjnldsv** [comment]
@tcitworld what is the status on this PR ? :)

**Pull request closed without merging**